ouroboros-consensus/test/consensus-test/Test/Consensus/MiniProtocol/ObjectDiffusion/Smoke.hs

```diff
@@ -0,0 +1,302 @@
+      ((\x -> "Inbound (Client): " ++ show x) `contramap` tracer)
+      codecObjectDiffusionId
+      inboundChannel
+      (objectDiffusionInboundPeerPipelined inbound)
+      >> pure ()
+
+--- The core logic of the smoke test is shared between the generic smoke tests for ObjectDiffusion, and the ones specialised to PerasCert/PerasVote diffusion
+prop_smoke_object_diffusion ::
+  ( Eq object
+  , Show object
+  , Ord objectId
+  , NoThunks objectId
+  , Show objectId
+  , NoThunks object
+  , Ord ticketNo
+  ) =>
+  ProtocolConstants ->
+  [object] ->
+  ( forall m.
+    IOLike m =>
+    ObjectDiffusionOutbound objectId object m () ->
+    Channel m (AnyMessage (ObjectDiffusion objectId object)) ->
+    (Tracer m String) ->
+    m ()
+  ) ->
+  ( forall m.
+    IOLike m =>
+    ObjectDiffusionInboundPipelined objectId object m () ->
+    (Channel m (AnyMessage (ObjectDiffusion objectId object))) ->
+    (Tracer m String) ->
+    m ()
+  ) ->
+  ( forall m.
+    IOLike m =>
+    m
+      ( ObjectPoolReader objectId object ticketNo m
+      , ObjectPoolWriter objectId object m
+      , m [object]
+      )
+  ) ->
+  Property
+prop_smoke_object_diffusion
+  (ProtocolConstants (maxFifoSize, maxIdsToReq, maxObjectsToReq))
+  objects
+  runOutboundPeer
+  runInboundPeer
+  mkPoolInterfaces =
+    let
+      simulationResult = runSimStrictShutdown $ do
+        let tracer = nullTracer
+
+        traceWith tracer "========== [ Starting ObjectDiffusion smoke test ] =========="
+        traceWith tracer (show objects)
+
+        (outboundPoolReader, inboundPoolWriter, getAllInboundPoolContent) <- mkPoolInterfaces
+        controlMessage <- uncheckedNewTVarM Continue
+
+        let
+          inbound =
+            objectDiffusionInbound
+              tracer
+              ( maxFifoSize
+              , maxIdsToReq
+              , maxObjectsToReq
+              )
+              inboundPoolWriter
+              nodeToNodeVersion
+              (readTVar controlMessage)
+
+          outbound =
+            objectDiffusionOutbound
+              tracer
+              maxFifoSize
+              outboundPoolReader
+              nodeToNodeVersion
+
+        withRegistry $ \reg -> do
+          (outboundChannel, inboundChannel) <- createConnectedChannels
+          outboundThread <-
+            forkLinkedThread reg "ObjectDiffusion Outbound peer thread" $
+              runOutboundPeer outbound outboundChannel tracer
+          inboundThread <-
+            forkLinkedThread reg "ObjectDiffusion Inbound peer thread" $
+              runInboundPeer inbound inboundChannel tracer
+          controlMessageThread <- forkLinkedThread reg "ObjectDiffusion Control thread" $ do
+            threadDelay 1000 -- give a head start to the other threads
+            atomically $ writeTVar controlMessage Terminate
+            threadDelay 1000 -- wait for the other threads to finish
+          waitAnyThread [outboundThread, inboundThread, controlMessageThread]
```

Comment:
Will this waiting condition give time to all the threads to send/receive all the objects?